default/python/AI/ProductionAI.py

```diff
@@ -436,17 +437,14 @@ def generate_production_orders():
                     continue
                 enrgy_shipyard_locs.setdefault(this_planet.systemID, []).append(pid)
                 if pid not in queued_building_locs and building_type.canBeProduced(empire.empireID, pid):
-                    res = fo.issueEnqueueBuildingProductionOrder(building_name, pid)
-                    debug(
-                        "Enqueueing %s at planet %s, with result %d"
-                        % (building_name, universe.getPlanet(pid).name, res)
-                    )
+                    res = building.enqueue(pid)
+                    debug("Enqueueing %s at planet %s, with result %d" % (building, universe.getPlanet(pid).name, res))
                     if res:
                         queued_building_locs.append(pid)
                         cost, time = empire.productionCostAndTime(production_queue[production_queue.size - 1])
                         building_expense += cost / time  # production_queue[production_queue.size -1].blocksize *
                         res = fo.issueRequeueProductionOrder(production_queue.size - 1, 0)  # move to front
-                        debug("Requeueing %s to front of build queue, with result %d", building_name, res)
+                        debug("Requeueing %s to front of build queue, with result %d", building, res)
```

Comment:
You could add an option at_front to BuildingType.enqueue and make it return a float, debugging should also be done there, then the whole block could be replaced by:
`building_expense += building.enqueue(pid, at_front=True)`

Reply:
Let's create a separate function for that in the production code